flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkContainersBuilder.java

```diff
@@ -0,0 +1,276 @@
+                            .setConfiguration(jobManagerConf)
+                            .setLogProperties(logProperties)
+                            .asJobManager()
+                            .build();
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to build JobManager image", e);
+        }
+        final GenericContainer<?> jobManager = buildContainer(jobManagerImage);
+        // Setup network for JobManager
+        jobManager
+                .withNetworkAliases(FlinkContainers.JOB_MANAGER_HOSTNAME)
+                .withExposedPorts(jobManagerConf.get(RestOptions.PORT));
+        // Setup logger
+        if (this.logger != null) {
+            jobManager.withLogConsumer(new Slf4jLogConsumer(this.logger).withPrefix("JobManager"));
+        }
+        return jobManager;
+    }
+
+    private List<GenericContainer<?>> buildTaskManagerContainers() {
+        List<GenericContainer<?>> taskManagers = new ArrayList<>();
+        for (int i = 0; i < numTaskManagers; i++) {
+            // Configure TaskManager
+            final Configuration taskManagerConf = new Configuration();
+            taskManagerConf.addAll(this.conf);
+            final String taskManagerHostName = FlinkContainers.TASK_MANAGER_HOSTNAME_PREFIX + i;
+            taskManagerConf.set(TaskManagerOptions.HOST, taskManagerHostName);
+            // Build TaskManager container
+            final ImageFromDockerfile taskManagerImage;
+            try {
+                taskManagerImage =
+                        new FlinkImageBuilder()
+                                .setConfiguration(taskManagerConf)
+                                .setLogProperties(logProperties)
+                                .asTaskManager()
+                                .build();
+            } catch (Exception e) {
+                throw new RuntimeException("Failed to build TaskManager image", e);
+            }
+            final GenericContainer<?> taskManager = buildContainer(taskManagerImage);
+            // Setup network for TaskManager
+            taskManager.withNetworkAliases(taskManagerHostName);
+            // Setup logger
+            if (this.logger != null) {
+                taskManager.withLogConsumer(
+                        new Slf4jLogConsumer(this.logger).withPrefix("TaskManager-" + i));
+            }
+            taskManagers.add(taskManager);
+        }
+        return taskManagers;
+    }
+
+    private GenericContainer<?> buildZookeeperContainer() {
+        final ImageFromDockerfile zookeeperImage;
+        try {
+            zookeeperImage =
+                    new FlinkImageBuilder()
+                            .setConfiguration(this.conf)
+                            .useCustomStartupCommand(
+                                    "flink/bin/zookeeper.sh start-foreground 1 && tail -f /dev/null")
+                            .setImageName("flink-zookeeper")
+                            .build();
```

Comment:
Nit: Can we extract this to its own class and extend the `GenericContainer` I think we should also use an official zookeeper image and put the  image name in `DockerImageVersions`. AFAIK Flink only supports a specific set of zookeeper versions.